Sources/ZIPFoundation/Data+Compression.swift

```diff
@@ -16,14 +16,51 @@ public typealias CRC32 = UInt32
 /// - Parameters:
 ///   - data: A chunk of `Data` to consume.
 /// - Throws: Can throw to indicate errors during data consumption.
-public typealias Consumer = (_ data: Data) throws -> Void
+public typealias Consumer<D: ContiguousBytes> = (_ data: D) throws -> Void
 /// A custom handler that receives a position and a size that can be used to provide data from an arbitrary source.
 /// - Parameters:
 ///   - position: The current read position.
 ///   - size: The size of the chunk to provide.
 /// - Returns: A chunk of `Data`.
 /// - Throws: Can throw to indicate errors in the data source.
-public typealias Provider = (_ position: Int, _ size: Int) throws -> Data
+public protocol Provider {
+    associatedtype SubSequence: ContiguousBytes
+
+    var count: Int { get }
+
+    func readData(position: Int, ofLength: Int) throws -> SubSequence
+}
+
+class PointerProvider: Provider {
+    typealias SubSequence = Data
+
+    let count: Int
+    let file: UnsafeMutablePointer<FILE>
+
+    init(size: Int, file: UnsafeMutablePointer<FILE>) {
+        self.count = size
+        self.file = file
+    }
+
+    func readData(position: Int, ofLength: Int) throws -> Data {
+        return try Data.readChunk(of: ofLength, from: file)
+    }
+}
+
+extension RandomAccessCollection where Self.Index == Int, Self.SubSequence: ContiguousBytes {
+    public func readData(position: Int, ofLength: Int) throws -> Self.SubSequence {
+        let endPos = (position + ofLength)
+        return self[position..<endPos]
+    }
+}
+extension Data: Provider {}
+extension Array: Provider where Self.Element == UInt8 {}
+extension ArraySlice: Provider where Self.Element == UInt8 {}
+extension ContiguousArray: Provider where Self.Element == UInt8 {}
+extension UnsafeBufferPointer: Provider where Self.Element == UInt8 {}
+extension UnsafeRawBufferPointer: Provider {}
+extension UnsafeMutableBufferPointer: Provider where Self.Element == UInt8 {}
+extension UnsafeMutableRawBufferPointer: Provider {}
```

Comment:
These types can now be passed in when a Provider is expected, without any wrapping.